### PR TITLE
[WICKET-7033] upload to a mounted resource

### PR DESCRIPTION
**Pull request by reiern70**
This PR adds:

- uploading file(s) into a mounted resource.
- Improve upload progress bar to work witth uploading to a mounted resource.

[Screencast from 2023-04-11 14-02-33.webm](https://user-images.githubusercontent.com/462655/231141442-cc82962e-131e-42f4-93a5-9fe145b366a5.webm)

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:59]: shouldn't this call be in the `else` clause ?

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:96]:
```suggestion
            Wicket.Log.log("Too late to cancel: the upload has already finished.");
```

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:93]:
```suggestion
            Wicket.Log.log("the upload has been canceled!");
```

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:97]: For easier debugging I'd suggest to use ` this.inputName ` in the log messages above.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FolderUploadsFileManager.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FolderUploadsFileManager.java:28]: javadoc

File: wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadApplication.java
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadApplication.java:66]: 100M is too big. We deploy the examples app at https://examples9x.wicket.apache.org and this may consume out disks quickly.

File: wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html:12]:
```suggestion
	<p>Wicket can upload to a <strong>mounted resource.</strong></p>
```

File: wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html:13]:
```suggestion
	<p>This upload is done via jQuery and does not block Wicket AJAX</p>
```

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/UploadProgressBar.java
Comment (martin-g) [inline wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/UploadProgressBar.java:134]: You can call `this(id, null, Args.nonNull(uploadField, "uploadField"))`
And move the non-null Form from https://github.com/apache/wicket/pull/571/files#diff-f75b5636ac8b2157552b9755243a10c168527c2132bd8ce61393386a49482047R183 to https://github.com/apache/wicket/pull/571/files#diff-f75b5636ac8b2157552b9755243a10c168527c2132bd8ce61393386a49482047R157

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js
Comment (martin-g) [inline wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js:140]: Why use `hidden` attribute ?
Can we use Wicket.DOM's `hide()` and `show()` instead ?

Comment (reiern70) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js:140]: Maybe. I need to check. I just modified this because we had some client side errors in some conner cases in our app.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java
Comment (reiern70) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java:122]: Maybe we should use i18n here...

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:384]: We will need them for our use case :-( I can always monkey pacth code but not nice...

**Comment by reiern70** [comment]
@martin-g Many thanks for all your comments. Addressing them ASAP

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadResourceReference.java
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadResourceReference.java:42]: You are right!

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:384]: Return `CharSequence` instead

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:316]: I got rid of this.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:93]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java:78]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java:55]: Done

File: wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html
Comment (reiern70) [reply wicket-examples/src/main/java/org/apache/wicket/examples/upload/UploadToResourcePage.html:12]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadResourceReference.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadResourceReference.java:39]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:116]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:48]: Done

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js
Comment (reiern70) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js:140]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:163]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java:196]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: Got rid of deleting. This we can do when needed in our application.

Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: > Have you tested with `../../../etc/passwd` ?

No.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: Fixed. Leftover of when thsi component only worked witha single fiel

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: Done.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: Done.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js]: Done

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.js:96]: Done.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/FileUploadToResourceField.java]: done

**Comment by reiern70** [comment]
@martin-g Many thanks again for your review. I have addressed ou comments. Could you do another review pelase?

**Review by bitstorm — CHANGES_REQUESTED** [review]
Once Maven build is happy I'm ok merging with 9.x . I don't see any reason to postpone it

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js
Comment (solomax) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js:65]: Just curious: why `var` is used and not `let` and `const`?

Comment (reiern70) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js:65]: > Just curious: why `var` is used and not `let` and `const`?

If you use let and const then the project will not build. I think there is some logic compiling and minimizing javascript and there are some settings for a target version of JavaScript that does not allow const and let. Only vars...

**Comment by reiern70** [comment]
> Once Maven build is happy I'm ok merging with 9.x . I don't see any reason to postpone it

Locally it builds.

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js
Comment (solomax) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js]: Understood, thanks for the clarifications!

Comment (reiern70) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/form/upload/progressbar.js]: > Understood, thanks for the clarifications!

Pleasure. Initialy I had those and **maven clean failed**. I juess this is to make sure our JavaScript works on older browsers too

**Comment by reiern70** [comment]
Fixed that warning,

**Comment by reiern70** [comment]
Squashed all commits and reabsed on top of 9.x + force pushed.

**Review by bitstorm — APPROVED** [review]

**Comment by reiern70** [comment]
@bitstorm Shall I wait till you cut release for merging? Or can I merge and you cout release from previosu commit?

**Comment by bitstorm** [comment]
> @bitstorm Shall I wait till you cut release for merging? Or can I merge and you cout release from previosu commit?

@reiern70  In the end I think it's better waiting after 9.13.0 is officially released. I will start the vote in few ours. Than I will merge your PR into 9.x branch.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java
Comment (solomax) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java:99]: FileCountLimitExceededException exception should also be handled,

Please see ade93326346e20587242b501422e43ae6937a843

Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java:99]: Ok. Then probably the resource needs a setting for maxFiles too

**Pull request merged**
